CRM/Utils/Cache/SqlGroup.php

```diff
@@ -96,11 +111,47 @@ public function __construct($config) {
    * @return bool
    */
   public function set($key, $value, $ttl = NULL) {
-    if ($ttl !== NULL) {
-      throw new \RuntimeException("FIXME: " . __CLASS__ . "::set() should support non-NULL TTL");
+    CRM_Utils_Cache::assertValidKey($key);
+
+    $lock = Civi::lockManager()->acquire("cache.{$this->group}_{$key}._null");
+    if (!$lock->isAcquired()) {
+      throw new \CRM_Utils_Cache_CacheException("SqlGroup: Failed to acquire lock on cache key.");
```

Comment:
Are these largely ignored? Failure to aquire a lock should not be a fatal error

Reply:
FWIW, this behavior is reproducing the behavior of `CRM_Core_BAO_Cache::setItem()`. The only change (based on feedback from Seamus) was replacing `CRM_Core_Error::fatal(...)` with `throw new Exception(...)`.

IIRC, what you're referring to about the hard-failure for CiviMail is addressed within the locking subsystem (ie within `CRM_Core_Lock`). That has some hacks which say, "If we're doing a CiviMail cronjob, then we won't try to acquire more locks, and we'll tell callers that our locks are working just fine." From the perspective of a specific lock user, they should still request/check locks as normal. The fudgery (as I understood it) was encapsulated within `CRM_Core_Lock`.

This code should behave just as well (or badly) as the previous code from `CRM_Core_BAO_Cache`.

Reply:
The only reason why i suggested Exception instead of fatal is that i believed that fatal calls were being removed in favor of exceptions.